python/python/bystro/supervised_ppca/gf_dropout_pt.py

```diff
@@ -0,0 +1,314 @@
+from bystro.supervised_ppca.gf_generative_pt import PPCA
+
+
+def _get_projection_matrix(W_, sigma_):
+    """
+    This is currently just implemented for PPCA due to nicer formula. Will
+    modify for broader application later.
+
+    Computes the parameters for p(S|X)
+
+    Description in future to be released paper
+
+    Parameters
+    ----------
+    W_ : pt.Tensor(n_components,p)
+        The loadings
+
+    sigma_ : pt.tensor
+        Isotropic noise
+
+    Returns
+    -------
+    Proj_X : pt.tensor(n_components,p)
+        Beta such that np.dot(Proj_X, X) = E[S|X]
+
+    Cov : pt.tensor(n_components,n_components)
+        Var(S|X)
+    """
+    n_components = int(W_.shape[0])
+    eye = torch.tensor(np.eye(n_components))
+    M = torch.matmul(W_, torch.transpose(W_, 0, 1)) + sigma_ * eye
+    Proj_X = torch.linalg.solve(M, W_)
+    Cov = torch.linalg.inv(M) * sigma_
+    return Proj_X, Cov
+
+
+class PPCADropout(PPCA):
+    """
+    This implements supervised PPCA according to the paper draft that I'm
+    working on. That is the generative mechanism matches probabilistic 
+    PCA with  isotropic variance. However, a variational lower bound on
+    a predictive objective is used to ensure that a subset of latent 
+    variables are predictive of an auxiliary task.
+
+    Parameters
+    ----------
+    n_components : int
+        The latent dimensionality
+
+    n_supervised : int
+        The number of predictive latent variables
+            
+    prior_options : dict
+        The hyperparameters for the prior on the parameters
+
+    mu : float>0,default=1.0
+
+    gamma : float,default=10.0
+
+    delta : 5.0
+        
+
+    """
+
+    def __init__(
+        self,
+        n_components=2,
+        n_supervised=1,
+        prior_options=None,
+        mu=1.0,
+        gamma=10.0,
+        delta=5.0,
+        training_options=None,
+    ):
+        self.mu = float(mu)
+        self.gamma = float(gamma)
+        self.delta = float(delta)
+        self.n_supervised = int(n_supervised)
+        super().__init__(
+            n_components=n_components,
+            prior_options=prior_options,
+            training_options=training_options,
+        )
+        self._initialize_save_losses()
+        self.losses_supervision = np.empty(
+            self.training_options["n_iterations"]
+        )
+        self.W_ = None
+        self.sigma2_ = None
+        self.B_ = None
+
+    def fit(self, X, y, task="classification", progress_bar=True, seed=2021):
+        """
+        Fits a model given covariates X as well as option labels y in the
+        supervised methods
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,n_covariates)
+            The data
+
+        y : np.array-like,(n_samples,n_prediction)
+            Covariates we wish to predict. For now lazy and assuming
+            logistic regression.
+
+        task : string,default='classification'
```

Comment:
it would be helpful to explicitly list the possible alternatives here (is the string "multinomial regression" a valid input?) and use a `match` statement to explicitly do the right thing for each possible value of `task` every time we reference it.  This line seems to disagree with the block ending on L176, which doesn't make any mention of multinomial regression.

You can have mypy spot these errors for you by defining the valid tasks as follows:

`Task = Literal["classification", "prediction", "multinomial_regression"]`

then annotating the input as type `Task` and ending each if-elif-block with an else with a call to assert_never():

```python
def assert_never(value: NoReturn) -> NoReturn:
    assert False, f"This code should never be reached, got: {value}"
```

Or you could do it by hand but the trade-off is that you have to worry about keeping everything in sync.

Reply:
In the next PR will add support for multinomial regression, but would be nice to leave in separate pr.